contrib/dracut/90zfs/module-setup.sh.in

```diff
@@ -85,7 +85,13 @@ install() {
 	fi
 
 	# Synchronize initramfs and system hostid
-	zgenhostid -o "${initdir}/etc/hostid" "$(hostid)"
+	if [ -e @sysconfdir@/hostid ]; then
+		inst @sysconfdir@/hostid
+		type mark_hostonly >/dev/null 2>&1 && mark_hostonly @sysconfdir@/hostid
+	elif HOSTID="$(hostid 2>/dev/null)" && [ "${HOSTID}" != "00000000" ]; then
+		zgenhostid -o "${initdir}@sysconfdir@/hostid" "${hostid}"
```

Comment:
I think `${hostid}` should be `${HOSTID}` here.

also I think it'd be better to check for zeroes first and resort to copying only if that condition is false and hostid file exists, so behaviour for non-musl systems does not change, besides marking hostonly.

consider alternative code
```bash
HOSTID="$(hostid)"
if [ "${HOSTID}" != "00000000" ]; then
    zgenhostid -o "${initdir}@sysconfdir@/hostid" "${HOSTID}"
    type mark_hostonly >/dev/null 2>&1 && mark_hostonly @sysconfdir@/hostid
else
	# probably musl stub gethostid() returned 0
	if [ -f @sysconfdir@/hostid ]; then
		inst @sysconfdir@/hostid
        type mark_hostonly >/dev/null 2>&1 && mark_hostonly @sysconfdir@/hostid
    fi
fi
```

Reply:
if file exists, hostid should already return it's value.

Reply:
You're right that `${hostid}` should be `${HOSTID}`. I also think that moving from `-e` to `-f` in the test is a good choice, although I opted for `-e` because the immediately preceding tests also use `-e`.

I think we should reduce the dependency on `hostid`, not elevate it. ZFS does not use the same mechanism for defining the hostid as the `hostid(1)` executable; it favors `spl_hostid` on the kernel command line and falls back to directly reading `/etc/hostid` (unless `spl` is configured to look elsewhere with another command-line parameter, but if that happens, this whole block is broken anyway). If `/etc/hostid` is directly a source of truth for ZFS, let's copy that truth without relying on an intermediary.

From a functional standpoint, the difference only matters if there exists a possibility that `hostid(1)` prints something *other than* the contents of `/etc/hostid`. I don't know whether this possibility exists.

If the `hostid` check is elevated, I still advocate the construct
```
if HOSTID="$(hostid 2>/dev/null)" && [ "${HOSTID}" != "00000000" ]
```
because it's probably a good idea to consider any stdout from a `hostid` that returns an error code as invalid.

Reply:
> I think we should reduce the dependency on hostid, not elevate it. 

I strongly agree with this point: the dracut module uses `hostid(1)` as if its behavior were guaranteed to be the same as what the kernel module (which reads `spl_hostid` from the kernel command line or the `/etc/hostid` file) and the ZFS user space tools (which read `/etc/hostid` directly as well) do. `hostid(1)` is an interface to libc's `gethostid(3)`, which, as a non-standard function, can be implemented in all sorts of ways. Moving to treating the hostid as close to what the other parts of ZFS do sounds to me like the best way forward.

> From a functional standpoint, the difference only matters if there exists a possibility that hostid(1) prints something other than the contents of /etc/hostid. I don't know whether this possibility exists.

This possibility does exist on glibc, where, if `/etc/hostid` can't be opened, `gethostid(3)` twiddles the current IP address into an `uint32_t` and returns that. This is a value that is not at all guaranteed to stay the same, and is miles away from what ZFS treats as the system's hostid.

Reply:
Everybody happy with the updated dracut change?

Reply:
The current state reflects my intended fix.

Reply:
yeah it's ok.
in gentoo we have another initrd generator which supports zfs (genkernel) and it copies hostid file just fine, instead of generating it.